### PR TITLE
Generate version-string from git-describe

### PR DESCRIPTION
**Pull request by stefan-koch-sociomantic**
- adds make target for `views/VERSION` which uses git describe to get the version number
- string imports `views/VERSION` if available and uses it as version-string
- `-debug` is added if in debug

### CODE REVIEW

File: views/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+*
+!.gitignore
```
Comment (leandro-lucarella-sociomantic) [inline views/.gitignore]: I don't think you need to specify `.gitignore` because AFAIK `*` doesn't match files starting with `.`.

In any case, it seems more sensible IMHO to just put `views/` in the top-level `.gitignore` instead.

BTW, what is this folder for? Why it exists and why should it be ignored, some clarification in the commit message would be helpful!

File: makefile [hunk 2]
```diff
@@ -1,7 +1,7 @@
 SRC := $(shell find src -name "*.d") \
 	$(shell find libdparse/src -name "*.d")
 INCLUDE_PATHS := -Ilibdparse/src -Isrc
-DMD_COMMON_FLAGS := -dip25 -w $(INCLUDE_PATHS)
+DMD_COMMON_FLAGS := -dip25 -w $(INCLUDE_PATHS) -Jviews
```
Comment (leandro-lucarella-sociomantic) [inline makefile]: Oh, I see now the "views" is for string inclusion. Why is it called "views" though? It looks like a funny name. But please squash the change adding the directory to the `.gitignore` file with this one.

File: makefile [hunk 3]
```diff
@@ -15,6 +15,9 @@ GDC ?= gdc
 
 dmd: bin/dfmt
 
+views/VERSION : $(SRC)
+	git describe --tags | cut -c2- | cut -f1-2 -d- > views/VERSION
```
Comment (leandro-lucarella-sociomantic) [inline makefile]: This needs a comment, what are you cutting and why? Also, use `> $@` instead of `> views/VERSION`, so if for some reason the file is renamed you don't need to update the script too.

Comment (leandro-lucarella-sociomantic) [reply makefile]: Also, this probably needs to get built unconditionally. If you change a README file and commit for example, the git hash /tag you are building will change, but the code will say you built another version. While the binary would be the same, it seems misleading.

You can unconditionally generate this file every time `make` runs via this trick:

```make
__dummy_version := $(shell git describe --tags | cut -c2- | cut -f1-2 -d- > views/VERSION)
```

File: makefile [hunk 4]
```diff
@@ -28,8 +31,9 @@ test: debug
 bin/dfmt-test: $(SRC)
 	$(DC) $(DMD_TEST_FLAGS) $^ -of$@
 
-bin/dfmt: $(SRC)
-	$(DC) $(DMD_FLAGS) $^ -of$@
+bin/dfmt: views/VERSION
+	$(DC) $(DMD_FLAGS) $(SRC) -of$@
```
Comment (leandro-lucarella-sociomantic) [inline makefile]: Better filter files: `$(DC) $(DMD_FLAGS) $(filter %.d,$^) -of$@`

File: makefile [hunk 5]
```diff
@@ -28,8 +31,9 @@ test: debug
...
+
+debug: views/VERSION
+	$(DC) $(DMD_DEBUG_FLAGS) $(SRC) -ofbin/dfmt
```
Comment (leandro-lucarella-sociomantic) [inline makefile]: Same.

File: makefile [hunk 6]
```diff
@@ -2,7 +2,7 @@ SRC := $(shell find src -name "*.d") \
 	$(shell find libdparse/src -name "*.d")
 INCLUDE_PATHS := -Ilibdparse/src -Isrc
 DMD_COMMON_FLAGS := -dip25 -w $(INCLUDE_PATHS) -Jviews
-DMD_DEBUG_FLAGS := -g $(DMD_COMMON_FLAGS)
+DMD_DEBUG_FLAGS := -debug -g $(DMD_COMMON_FLAGS)
```
Comment (leandro-lucarella-sociomantic) [inline makefile]: Please do this in a separate commit.

File: src/dfmt/main.d [hunk 7]
```diff
@@ -5,7 +5,25 @@
...
+
+static if (is(typeof(import("VERSION"))))
+{
+    import std.string : chomp;
+    private enum VERSION = chomp(import("VERSION")) ~ DEBUG_SUFFIX;
+}
+else
+{
+    private enum VERSION = "0.6.x" ~ DEBUG_SUFFIX;
+}
```
Comment (leandro-lucarella-sociomantic) [inline src/dfmt/main.d]: I think this should fail if `VERSION` can't be found. This will get outdated extremely soon otherwise, what's the point on keeping a version like this?

File: views/.gitignore (hunk 1 above)
Comment (stefan-koch-sociomantic) [reply views/.gitignore]: I need the .gitigonre because git ignores empty directories.
I cannot add views if it has no files.
And views need to exist for `-J` to work

Comment (leandro-lucarella-sociomantic) [reply views/.gitignore]:
```make
__dummy_version := $(shell mkdir -p views && git describe --tags | cut -c2- | cut -f1-2 -d- > views/VERSION)
```

Comment (stefan-koch-sociomantic) [reply views/.gitignore]: I don't like this :-(

File: makefile (hunk 2 above)
Comment (stefan-koch-sociomantic) [reply makefile]: `views` is the default folder dub uses for string inclusion

File: views/.gitignore (hunk 1 above)
Comment (leandro-lucarella-sociomantic) [reply views/.gitignore]: Why?

File: makefile (hunk 2 above)
Comment (leandro-lucarella-sociomantic) [reply makefile]: OK then. I wonder where does that comes from, but if it's some sort of defacto standard, I'm fine with it.

File: makefile (hunk 4 above)
Comment (stefan-koch-sociomantic) [reply makefile]: `$^` are the dependencies right ? in this case `views/VERSION`
that would not work

File: makefile (hunk 2 above)
Comment (stefan-koch-sociomantic) [reply makefile]: It comes from the fact that vibed's `diet-templates` are in there.
which are the view in M-V-C.

Comment (leandro-lucarella-sociomantic) [reply makefile]: Typical "historical reasons" :man_facepalming:

File: makefile (hunk 4 above)
Comment (leandro-lucarella-sociomantic) [reply makefile]: `$^` are all the dependencies, yes, but `$(filter %.d,$^)` are only the dependencies matching `*.d` so to say, which is exactly what you want to pass to DMD, so it will work, it does for Makd :)

Comment (stefan-koch-sociomantic) [reply makefile]: this expands to
`dmd -O -inline -dip25 -w -Ilibdparse/src -Isrc -Jviews  -ofbin/dfmt`
since views/VERSION is the only dependency

File: views/.gitignore (hunk 1 above)
Comment (stefan-koch-sociomantic) [reply views/.gitignore]: it looks hacky :)

File: src/dfmt/main.d (hunk 7 above)
Comment (stefan-koch-sociomantic) [reply src/dfmt/main.d]: make is not the only build-system used;
The dub path has to work as well; I cannot figure out howto make writing of the version file work on windows.

**Comment by stefan-koch-sociomantic** [comment]
The travis error comes from dmd-nightly not being downloaded.

**Comment by stefan-koch-sociomantic** [comment]
@leandro-lucarella-sociomantic please have another look

File: makefile (hunk 6 above)
Comment (leandro-lucarella-sociomantic) [reply makefile]: NOT addressed yet.

File: src/dfmt/main.d [hunk 8]
```diff
@@ -4,8 +4,61 @@
...
+        if (minusCount > 1)
+		result = null;
+
+        return result ? result ~ DEBUG_SUFFIX
+                      : gitDescribeOutput ~ DEBUG_SUFFIX;
+
+    }
+    else
+    {
+        return "0.6.x" ~ DEBUG_SUFFIX;
```
Comment (leandro-lucarella-sociomantic) [inline src/dfmt/main.d]: I don't recall why is this still there. Can't you make `git describe > views/VERSION` work for all platform so no manual version has to be maintained?

**Comment by leandro-lucarella-sociomantic** [comment]
```
$ source "$(CURL_USER_AGENT="$CURL_USER_AGENT" bash install.sh dmd-nightly --activate)"
Warning: No gpg tool found to verify downloads.
curl: (56) Recv failure: Operation timed out
/Users/travis/.travis/job_stages: line 57: : No such file or directory
```

I'll re-run :crossed_fingers:
